File: Problems/Algorithms/225. Implement Stack using Queues/Stack.kt
class MyStack() {
    
    private val q1: ArrayDeque<Int>
    private val q2: ArrayDeque<Int>
    private var top: Int
    
    init {
        q1 = ArrayDeque()
        q2 = ArrayDeque()
        top = 0
    }

    fun push(x: Int) {
        q1.add(x)
        top = x
    }

    fun pop(): Int {
        while (q1.size > 1) {
            top = q1.removeFirst()
            q2.add(top)
        }
        return q1.removeFirst().also {
            while (!q2.isEmpty()) {
                q1.add(q2.removeFirst())
            }
        }
    }

    fun top(): Int {
        return top
    }

    fun empty(): Boolean {
        return q1.isEmpty()
    }

}

/**
 * Your MyStack object will be instantiated and called as such:
 * var obj = MyStack()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.empty()
 */

File: Problems/Algorithms/225. Implement Stack using Queues/stack.cpp
#include <queue>

using namespace std;

class MyStack {
public:
    queue<int> queue1, queue2;

    MyStack() {}
    
    void push(int x) {
        while (queue1.size() > 0) {
            queue2.push(queue1.front());
            queue1.pop();
        }

        queue1.push(x);
        while (queue2.size() > 0) {
            queue1.push(queue2.front());
            queue2.pop();
        }
    }
    
    int pop() {
        int peek = queue1.front();
        queue1.pop();

        return peek;
    }
    
    int top() {
        return queue1.front();
    }
    
    bool empty() {
        return queue1.size() == 0;
    }
};

/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack* obj = new MyStack();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->top();
 * bool param_4 = obj->empty();
 */